pingcap/tidb
# *: add privilege check for slow_query and cluster memory table

### What problem does this PR solve? 
Before this PR, TiDB doesn't have a privilege check on below tables:
* `SLOW_QUERY`
* `CLUSTER_SLOW_QUERY`
* `CLUSTER_PROCESSLIST`

Mainly, there is 2 points need to do:

**First**

For doing privilege checks in the cluster table, we need to add `user` and `host` info to RPC request. It needs to change the `tipb` protobuf, see PR: https://github.com/pingcap/tipb/pull/169

**Second**

What kind of privilege is needed for query `SLOW_QUERY`?  I think the information in `SLOW_QUERY` and `PROCESSLIST` is the same, So I use the same `PROCESS` privilege for checking `SLOW_QUERY` and `CLUSTER_SLOW_QUERY` privilege.


eg:
```sql
-- For user root: 
mysql -u root -h 127.0.0.1 -P 4000
tidb>select user,count(*) from `CLUSTER_SLOW_QUERY` group by user;
+------+----------+
| user | count(*) |
+------+----------+
| test | 2        |
| root | 95622    |
|      | 743      |       -- name is "" means it is a internal sql.
+------+----------+
tidb>select * from `CLUSTER_PROCESSLIST`;
+---------------+----+------+-----------+--------------------+---------+------+-------+-------------------------------------+-----+----------------------------------------+
| ADDRESS       | ID | USER | HOST      | DB                 | COMMAND | TIME | STATE | INFO                                | MEM | TxnStart                               |
+---------------+----+------+-----------+--------------------+---------+------+-------+-------------------------------------+-----+----------------------------------------+
| 0.0.0.0:10080 | 4  | test | 127.0.0.1 | information_schema | Sleep   | 131  | 2     | <null>                              | 0   |                                        |
| 0.0.0.0:10080 | 8  | test | 127.0.0.1 | <null>             | Query   | 40   | 2     | select sleep(100)                   | 0   |                                        |
| 0.0.0.0:10080 | 10 | root | 127.0.0.1 | <null>             | Query   | 27   | 2     | select sleep(100)                   | 0   |                                        |
| 0.0.0.0:10080 | 1  | root | 127.0.0.1 | information_schema | Query   | 0    | 2     | select * from `CLUSTER_PROCESSLIST` | 0   | 01-11 15:02:32.451(413853588508114945) |
+---------------+----+------+-----------+--------------------+---------+------+-------+-------------------------------------+-----+----------------------------------------+

-- At the same time:
-- For user test:
mysql -u test -h 127.0.0.1 -P 4000
tidb>select user,count(*) from `CLUSTER_SLOW_QUERY` group by user;
+------+----------+
| user | count(*) |
+------+----------+
| test | 2        |
+------+----------+
tidb>select * from `CLUSTER_PROCESSLIST`;
+---------------+----+------+-----------+--------------------+---------+------+-------+-------------------------------------+-----+----------------------------------------+
| ADDRESS       | ID | USER | HOST      | DB                 | COMMAND | TIME | STATE | INFO                                | MEM | TxnStart                               |
+---------------+----+------+-----------+--------------------+---------+------+-------+-------------------------------------+-----+----------------------------------------+
| 0.0.0.0:10080 | 8  | test | 127.0.0.1 | <null>             | Query   | 62   | 2     | select sleep(100)                   | 0   |                                        |
| 0.0.0.0:10080 | 4  | test | 127.0.0.1 | information_schema | Query   | 0    | 2     | select * from `CLUSTER_PROCESSLIST` | 0   | 01-11 15:02:54.450(413853594275020801) |
+---------------+----+------+-----------+--------------------+---------+------+-------+-------------------------------------+-----+----------------------------------------+
```

### What is changed and how it works?


### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change

Side effects


Related changes

 - Need to cherry-pick to the release branch

Release note

 - Add privilege check for slow_query.
